cluster-autoscaler/core/scale_up.go

```diff
@@ -444,11 +456,10 @@ func ScaleUp(context *context.AutoscalingContext, processors *ca_processors.Auto
 			if err != nil {
 				return nil, err
 			}
+
 			// Node group id may change when we create node group and we need to update
 			// our data structures.
 			if oldId != bestOption.NodeGroup.Id() {
-				podsPassingPredicates[bestOption.NodeGroup.Id()] = podsPassingPredicates[oldId]
-				delete(podsPassingPredicates, oldId)
```

Comment:
IIUC, this means that for an autoprovisioned node group created with different id, we'll:
- run all predicates again (call computeCache(newId)),
- use the same nodeInfo object as in simulation.

Is this WAI? I'm OK with either assuming the result of simulation is fine, or with running predicates again on a node info built from actual instance template, but right now it seems we'll be running the same simulation twice for selected node group.

Reply:
Running same simulation was my intention. IIUC the previous code was just moving `podsPassingPredicates[oldId]` to `podsPassingPredicates[newId]`. I wanted to avoid explicit cache modifications from external code, at cost of one extra computation.

Reply:
Ack, let's leave it for now then.

Reply:
True, but we assume this in so many places that fixing it would probably require rethinking this whole file. Accessor won't help with e.g. cache invalidation - ideally when we replace or remove a nodeInfo for a node group, we should clean all results (errors, pods passing, pods not passing etc.) for this group.

Reply:
> True, but we assume this in so many places that fixing it would probably require rethinking this whole file.

Well - we mostly pass it around as a parameter. Which IMO is fine. The difference with caching is that it is stored within closure returned from `getPodsPredicatePassingCheckFunctions`. This is somewhat different and less readable IMO.

> Accessor won't help with e.g. cache invalidation - ideally when we replace or remove a nodeInfo for a node group, we should clean all results (errors, pods passing, pods not passing etc.) for this group.

This cache is rebuilt every iteration. I do not see why we would like to invalidate any entries in it really.

Reply:
> This is somewhat different and less readable IMO.

Agreed, but I don't think an accessor (= same logic, but yet another layer of abstraction to follow) would help much with it. I was thinking about encapsulating this in a "simulation results cache" object, and letting it own nodeInfos, but that's probably impractical given that they're used extensively by other part of the code.

> This cache is rebuilt every iteration. I do not see why we would like to invalidate any entries in it really.

If we create a node group, we may want to (eventually) run the simulation on the node built from actual instance template. But it's just a possibility at this point, yeah. Right now, we're only adding a new nodeInfo object (or rather, moving an existing one under a new key), and re-running simulations on it to update remaining cache objects. How about instead add another function, e.g. "addNodeGroupAlias(oldId, newId)", which will basically copy all the results associated with oldId to be associated with newId?